src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -508,13 +524,13 @@ __attribute__ ((visibility("default")))
 CUresult CUDAAPI cuMemFree(CUdeviceptr dptr)
 {
     CUresult result;
-    if (!gpu_initialized) {
-        int ret = MPL_gpu_init(0);
-        assert(ret == 0);
-    }
+    MPL_initlock_lock(&free_hook_mutex);
+    gpu_mem_hook_init();
```

Comment:
Is it good to call `gpu_mem_hook_init()` multiple times? It calls `dlopen`, potentially it has a nonnegligible delay. How about
```
if (!sys_cudaFree) {
    gpu_mem_hook_init();
}
```
?

Reply:
Works for me. I added it as a separate commit before the thread-safety addition.